subgraphs/minimal-legacy-exchange/src/pricing.ts

```diff
@@ -0,0 +1,137 @@
+import { BigDecimal } from '@graphprotocol/graph-ts'
+import {
+  BIG_DECIMAL_ONE,
+  BIG_DECIMAL_ZERO,
+  FACTORY_ADDRESS,
+  MINIMUM_NATIVE_LIQUIDITY,
+  NATIVE_ADDRESS,
+  STABLE_POOL_ADDRESSES,
+  STABLE_TOKEN_ADDRESSES,
+} from './constants'
+import { getOrCreateToken } from './functions'
+import { getTokenKpi } from './functions/token-data'
+import { Pool, Token, TokenKpi } from '../generated/schema'
+import { Factory as FactoryContract } from '../generated/templates/Pair/Factory'
+import { getPoolKpi } from './functions/pool-data'
+
+export const factoryContract = FactoryContract.bind(FACTORY_ADDRESS)
+
+export function getNativePriceInUSD(): BigDecimal {
+  let count = 0
+  let weightdPrice = BigDecimal.fromString('0')
+  let nativeReserve = BigDecimal.fromString('0')
+  let stablePrices: BigDecimal[] = []
+  let nativeReserves: BigDecimal[] = []
+
+  for (let i = 0; i < STABLE_POOL_ADDRESSES.length; i++) {
+    const address = STABLE_POOL_ADDRESSES[i]
+
+    const stablePool = Pool.load(address)
+    if (stablePool === null) {
+      continue
+    }
+    const stablePoolKpi = getPoolKpi(address)
+
+    if (
+      (stablePool.token0 == NATIVE_ADDRESS && stablePoolKpi.reserve0.lt(MINIMUM_NATIVE_LIQUIDITY)) ||
+      (stablePool.token1 == NATIVE_ADDRESS && stablePoolKpi.reserve1.lt(MINIMUM_NATIVE_LIQUIDITY))
+    ) {
+      continue
+    }
+
+    const stableFirst = STABLE_TOKEN_ADDRESSES.includes(stablePool.token0)
+
+    nativeReserve = nativeReserve.plus(!stableFirst ? stablePoolKpi.reserve0 : stablePoolKpi.reserve1)
+
+    nativeReserves.push(!stableFirst ? stablePoolKpi.reserve0 : stablePoolKpi.reserve1)
+
+    stablePrices.push(stableFirst ? stablePoolKpi.token0Price : stablePoolKpi.token1Price)
+
+    count = count + 1
+  }
+
+  if (count > 0) {
+    for (let j = 0; j < count; j++) {
+      const price = stablePrices[j]
+      const weight = nativeReserves[j].div(nativeReserve)
+      weightdPrice = weightdPrice.plus(price.times(weight))
+    }
+  }
+
+  return weightdPrice
+}
+
+/**
+ * Updates the token KPI price for the given token.
+ * Find the pool that contains the most liquidity and is safe from circular price dependency,
+ * (e.g. if DAI is priced off USDC, then USDC cannot be priced off DAI)
+ * @param tokenAddress The address of the token kpi to update
+ * @returns
```

Comment:
Take a look at the schema, I've stripped the schema to only contain the fields needed to do pricing

I've renamed TokenKpi.`whitelistedPools` to just `pools` and price anything that passes the minimum liquidity check. To get around the circular price dependency we have seen in other subgraphs, I've added a field to the `TokenKpi` entity, called `priceOffToken`.